orc8r/cloud/go/service/middleware/unary/identity_decorator.go

```diff
@@ -102,82 +101,113 @@ func SetIdentityFromContext(ctx context.Context, _ interface{}, info *grpc.Unary
 
 	ctxMetadata, ok := metadata.FromIncomingContext(ctx)
 	if !ok || ctxMetadata == nil {
-		glog.Info(ERROR_MSG_NO_METADATA)
 		// Metadata should always be present for GRPC client calls
 		// If we want to enable ANY calls from local clients, we need to
 		// change the return statement to:
 		//   return newCtx, newReq, resp, ensureLocalPeer(ctx)
-		// but, it may present a se3curity risk
-		err = status.Error(codes.Unauthenticated, ERROR_MSG_NO_METADATA)
-		return newCtx, newReq, resp, err
+		// but, it may present a security risk
+		return newCtx, newReq, resp, noMetadata()
 	}
 
-	// First, try to find the caller's identity
 	snlist, snok := ctxMetadata[CLIENT_CERT_SN_KEY]
-	if snok { // there is a certificate serial number (CSN) list in CTX
-		if len(snlist) != 1 {
-			// there can be only one CSN, error out if not
-			glog.Infof("Multiple CSNs found in metadata: %+v", ctxMetadata)
-			err = status.Error(codes.Unauthenticated, "Multiple CSNs present")
-		} else {
-			// One CSN is found, find Identity associated with it
-			var gwIdentity *protos.Identity
-			var certExpTime int64
-			// Check if SN is the reserved value used for all inter-orc8r calls
-			if snlist[0] == registry.ORC8R_CLIENT_CERT_VALUE {
-				return newCtx, newReq, resp, nil
-			}
-			gwIdentity, certExpTime, err = findGatewayIdentity(ctx, snlist[0], ctxMetadata)
-			if err == nil {
-				// If a valid GW Identity is found, add it into CTX for use
-				// by the callee
-				newCtx = gwIdentity.NewContextWithIdentity(protos.NewContextWithCertExpiration(ctx, certExpTime))
-				return newCtx, newReq, resp, err
-			}
-		}
-	} else {
-		// No CSNs found, check there is also no Certificate Common Names
-		if _, ok = ctxMetadata[CLIENT_CERT_CN_KEY]; ok {
-			// CN header is present while SN header is missing - possible
-			// security hack, either both or neither of the headers should be
-			// set
-			glog.Infof("CCN is present without SCN in metadata: %+v", ctxMetadata)
-			err = status.Error(codes.Unauthenticated, "Inconsistent Request Signature")
+	if snok {
+		newCtx, err = findIdentityFromCSNlist(snlist, ctx, ctxMetadata)
+		if err == nil {
+			return newCtx, newReq, resp, nil
 		}
+	} else if _, ok = ctxMetadata[CLIENT_CERT_CN_KEY]; ok {
+		// CN header is present while SN header is missing - possible
+		// security hack, either both or neither of the headers should be
+		// set
+		err = unexpectedCCN(ctxMetadata)
 	}
 
-	if info != nil {
-		// Check if the call is for a allowlisted method - anything is allowed
-		// do this check past possible identity decoration to still allow to add
-		// valid identity even to allowlisted requests
-		if _, ok := identityDecoratorBypassList[info.FullMethod]; ok {
-			// Bypass method (Bootstrapper & Co.), shortcut...
-			return newCtx, newReq, resp, nil
-		}
+	if isAllowlistedCall(info) {
+		return newCtx, newReq, resp, nil
 	}
 
-	// Only allow local clients if there is no previous errors (snok == false)
 	if err == nil {
```

Comment:
is the `err==nil` case reachable? in https://github.com/magma/magma/blob/830bf9beee8fc4b59351ad942d6ac4701bbaa1b2/orc8r/cloud/go/service/middleware/unary/identity_decorator.go#L113-L123 it looks like that both branches end on a non-nil `err`:

- if `err==nil` in the first branch, it is returned early
- `unexpectedCCN` never returns `nil`

Reply:
Yes, when `ok` is false in the `lese if` bracket, which is the case when both CSN and CNN are missing. It is a bit cryptic to see that, I tried refactoring a bit more and reversing the last cases (i.e. asking for `!ok` instead of `ok`) to make it more clear.

Reply:
ah, `else if` and not `else` :)